[tools/depends] enable asm for darwinembedded platforms - openssl, nettle, gnutls - update gaspreprocessor

fuzzard opened the pull request:
## Description
Enables building aarch64 assembly for darwinembedded platforms libs

- openssl
- nettle
- gnutls

Fix
- gas-preprocessor.pl - Update to libav github repo version. Add VERSION to track versioning for this in the future.

## Motivation and Context
Now that ios host triplets have been corrected, we can easily build aarch64 asm for darwinembedded platforms. Might as well make use of it.

## How Has This Been Tested?
Built and runtime tested on iOS/tvOS

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

fuzzard committed:
[depends/target] Darwinembedded gnutls enable building asm

Builds asm for aarch64 darwinembedded platforms

fuzzard committed:
[depends/target] darwinembedded nettle enable asm

enable building asm for aarch64 darwinembedded platforms

fuzzard committed:
[depends/target] darwinembedded openssl enable asm

enable building asm for aarch64 darwinembedded platforms

phunkyfish commented on tools/depends/target/gnutls/Makefile:
```diff
@@ -28,6 +24,10 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 	cd $(PLATFORM); patch -p1 -i ../size-max.patch
 	cd $(PLATFORM); patch -p1 -i ../add-dl-as-private-lib.patch
+ifeq ($(OS),darwin_embedded)
+	# Apple CCAS requires stripping ; from comment block # to build successfully
```
This is should be upstreamed right?

fuzzard replied:
I honestly dont think it would be accepted. It only affects ios/tvos build tools (not osx), and the fault is incorrectly processing a comment block. For whatever reason, its still trying to process control elements inside the comment.

In addition its a license header comment block, thats in every other source file in the lib. I cant see a gnu project changing a license header for the sake of a closed, niche platform.

phunkyfish replied:
Fair point.

fuzzard replied:
Turns out the issue stemmed from gas-preprocessor. Added commit that fixes that.
Not sure how that effects your ffmpegdirect stuff @phunkyfish

phunkyfish commented on tools/depends/configure.ac:
```diff
@@ -418,7 +418,7 @@ case $host in
     platform_includes="-isysroot $use_sdk_path"
     deps_dir="${sdk_name}_${use_cpu}-target-${build_type}"
-    AC_CHECK_LIB([z], [main], has_zlib=1, AC_MSG_WARN("No zlib support in toolchain. Will build libz."); has_zlib=0)
+    AC_SEARCH_LIBS([deflate],z, link_zlib=$ac_cv_search_zlib_main, link_zlib=-lz; AC_MSG_WARN("No zlib support in toolchain. Will build zlib."); has_zlib=1)
```
Can you explain this?

fuzzard replied:
absolutely may be a better way to do this. Im just not sure.

Context is, we need to add -lz to linker flags (in particular for this use case CMAKE_EXE_LINKER_FLAGS for use by mariadb-connector-c). 
AC_SEARCH_LIBS appears to append to $LIBS as per https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Libraries.html

So essentially, checking for deflate function in the z library, 3rd option is on found, as its pre-depends, its actually not found for ios/tvos, so this argument is really null anyway. last argument is on failure to find. I think ive got the has_zlib=1 wrong, i think it should be 0, i'll have to double check that. Either way, this adds -lz to $LIBS, which is then added to platform_ldflags at https://github.com/xbmc/xbmc/blob/c4405521ac0f1f8efacda83d4b0b2cf21bc7a7cc/tools/depends/configure.ac#L635

Realistically, i dont know the history behind the usage of ac_check_lib (again, it will always do the fail case of 4th argument anyway for ios/tvos), so we may just be able to add -lz explicitly to platform_ldflags and call it a day. Maybe need to set has_zlib=0, not totally sure where that gets used past configure

phunkyfish replied:
Ok, that explains why we don’t need this in a binary addon as we explicitly build the dependencies.

fuzzard replied:
yeah, thats the funny thing, we do in core as well, haha. i think the usage is totally redundant

phunkyfish replied:
Is their no package config file for Mariadb connector c?

fuzzard replied:
there is, just doesnt seem to link whats needed though.

phunkyfish replied:
Why not patch the PC file instead?

fuzzard replied:
And then the next dependency that needs it as well? just easier to set it globally imo.

fuzzard replied:
my understanding was a little flawed from my earlier comment.
I've fixed it up to be more accurate.

Essentially on darwin platforms, we wont actually build/use target zlib, as system zlib (10.14 with xcode 10.3 is 1.2.11 which is what we currently use in target depends).
The search finds the toolchain zlib, so we still add the linker flag using ac_search_libs. We could patch the pc file, but i think this is more robust long term, without carrying/maintaining a patch.

phunkyfish commented:
BTW, I have added this to ffmpegdirect.

fuzzard commented:
Theres a few other dependencies that support aarch64 asm, but they werent as simple to get running, and require some larger patching, so i havent bothered getting them running as yet. Baby steps ;)

fuzzard commented:
spent some more time, and actually found the gnutls issue i was patching was due to gas-preprocessor. Its regex on the input wasnt matching the comment block. 

I've altered the regex to work for this case, with what i believe will be minimal impact.
As there is no versioning for gas-preprocessor.pl, ive really no idea where it was originally sourced from. At a guess im assuming it stems from https://github.com/libav/gas-preprocessor but really not sure.
So not sure about upstreaming.

phunkyfish commented:
Ha, funny you went this route so. Cause we wouldn't have used the gas-preprocessor for anything other than ffmpeg. So I'll stick with V1 in the addon!

fuzzard committed:
[depends/native] native gas-preprocessor bump

update gas-preprocessor to version from libav. Add version file to attempt to allow
tracking of versioning in the future.

fuzzard commented:
Figure we just update gas-preprocessor completely. Ive pulled it from the libav github, and have added a version file with some details. Hoping this will allow us to track and manage any future updates/changes upstream a little easier.

Rechi commented on tools/depends/native/gas-preprocessor/README:
```diff
@@ -17,4 +10,4 @@ lipo -create -arch armv6 armv6/libavcodec/libavcodec.a -arch armv7 armv7/libavco
 and similar for each library. Then in XCode, make sure to build for both armv6
 and armv7. If you only care about one generation (since the armv6 devices are
-too slow for instance), then lipo is unnecessary of course.
+too slow for instance), then lipo is unnecessary of course.
```
add a new line at the end of file

Rechi commented on tools/depends/native/gas-preprocessor/VERSION:
```diff
@@ -0,0 +1,4 @@
+LIBNAME=gas-preprocessor
+BASE_URL=https://github.com/libav/gas-preprocessor
+COMMIT=d09971fad329d32df19f5bbafe88cf2f0ed04ed7
+COMMIT_DATE=2019/10/04
```
add a new line at the end of file

Rechi commented on tools/depends/target/mariadb/02-symbol_rename-gnutls_clash.patch:
```diff
@@ -0,0 +1,97 @@
+ ******************************************************************************/
+ 
+-my_bool hash_delete(HASH *hash,uchar *record)
++my_bool hash_delete1(HASH *hash,uchar *record)
```
Prefix the new name of the function with the library name instead of appending 1.

fuzzard replied:
done as requested. Used a define to minimise the patch as well. Hope that was what you meant in your other comment.

Rechi commented on tools/depends/target/mariadb/Makefile:
```diff
@@ -16,8 +24,8 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 	rm -rf $(PLATFORM); mkdir -p $(PLATFORM)/build
 	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
 	cd $(PLATFORM); patch -p1 -i ../01-android.patch
-	sed -ie 's| "DYNAMIC" | "STATIC" |' $(PLATFORM)/cmake/plugins.cmake
```
Why remove this?

fuzzard replied:
Isn't actually valid anymore

Rechi commented on tools/depends/target/mariadb/Makefile:
```diff
@@ -16,8 +24,8 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 	cd $(PLATFORM); patch -p1 -i ../01-android.patch
-	sed -ie 's| "DYNAMIC" | "STATIC" |' $(PLATFORM)/cmake/plugins.cmake
-	cd $(PLATFORM)/build; $(CMAKE) -DAUTH_GSSAPI=OFF -DWITH_UNITTEST:BOOL=OFF -DWITH_EXTERNAL_ZLIB:BOOL=ON -DWITH_CURL:BOOL=OFF ..
+	cd $(PLATFORM); patch -p1 -i ../02-symbol_rename-gnutls_clash.patch
```
Rename the function by using a preprocessor defines. Otherwise you always have to recheck the patch file.

fuzzard replied:
Hope this is what you meant with the updated patch. Seemed to work and be cleaner locally. If not, if you wouldnt mind elaborating a little.

Rechi commented on tools/depends/target/mariadb/Makefile:
```diff
@@ -16,8 +24,8 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
-	sed -ie 's| "DYNAMIC" | "STATIC" |' $(PLATFORM)/cmake/plugins.cmake
-	cd $(PLATFORM)/build; $(CMAKE) -DAUTH_GSSAPI=OFF -DWITH_UNITTEST:BOOL=OFF -DWITH_EXTERNAL_ZLIB:BOOL=ON -DWITH_CURL:BOOL=OFF ..
+	cd $(PLATFORM); patch -p1 -i ../02-symbol_rename-gnutls_clash.patch
+	cd $(PLATFORM)/build; $(CMAKE) -DWITH_SSL=ON ${CONFIGURE_FLAGS} -DWITH_UNIT_TESTS:BOOL=OFF -DWITH_EXTERNAL_ZLIB:BOOL=ON -DWITH_CURL:BOOL=OFF ..
```
Why not name the value of CONFIGURE_FLAGS variable explicitly here?

fuzzard replied:
Hope this is what you were thinking.

Rechi reviewed (changes requested):
Commit order seems to be wrong too.

Rechi commented on tools/depends/configure.ac:
```diff
@@ -418,7 +418,7 @@ case $host in
     platform_includes="-isysroot $use_sdk_path"
     deps_dir="${sdk_name}_${use_cpu}-target-${build_type}"
-    AC_CHECK_LIB([z], [main], has_zlib=1, AC_MSG_WARN("No zlib support in toolchain. Will build libz."); has_zlib=0)
+    AC_SEARCH_LIBS([deflate],z, has_zlib=1, AC_MSG_WARN("No zlib support in toolchain. Will build libz."); has_zlib=0)
```
This is wrong. As mariadb has to link zlib (because of openssl), it has to be fixed in the mariadb package.

fuzzard replied:
is there a reason not to have it as a linker flag for everything? We dont build zlib for apple platforms, as they have their own system version.

phunkyfish replied:
@Rechi any further comment here?

Rechi replied:
Because once the library which required it, doesn't need it anymore or gets remove this will be kept for no reason.

fuzzard replied:
It's used in like 7 other dependencies. But point taken. I'll change as requested.

Rechi reviewed (changes requested):
I don't see a reason for the version bumping of gnutls split across two commits.

fuzzard commented on tools/depends/target/mariadb/Makefile:
```diff
@@ -16,8 +24,8 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
-	sed -ie 's| "DYNAMIC" | "STATIC" |' $(PLATFORM)/cmake/plugins.cmake
-	cd $(PLATFORM)/build; $(CMAKE) -DAUTH_GSSAPI=OFF -DWITH_UNITTEST:BOOL=OFF -DWITH_EXTERNAL_ZLIB:BOOL=ON -DWITH_CURL:BOOL=OFF ..
+	cd $(PLATFORM); patch -p1 -i ../02-symbol_rename-gnutls_clash.patch
+	cd $(PLATFORM)/build; $(CMAKE) -DWITH_SSL=ON ${PLUGIN_BUILD_FLAGS} -DWITH_UNIT_TESTS:BOOL=OFF -DWITH_EXTERNAL_ZLIB:BOOL=ON -DWITH_CURL:BOOL=OFF ..
```
My reasoing was it made it a bit more clear what those values were for. Having them all inline, would you ever know what exactly they are, and why one is off?

fuzzard replied:
as for the patch comment, could you elaborate more on what that would look like?
Im quite certain the includes will happen appropriately, but im happy to go with whatever other method you are proposing, i just dont know how to implement that sorry.

Rechi replied:
just can add the comment here

fuzzard commented:
> I don't see a reason for the version bumping of gnutls split across two commits.

Sole reason was i can, and have only runtime tested on ios/osx. If i leave the split, a rollback for other platforms is a possibility, whilst still allowing ios/tvos to run the newer/tested version, with the asm acceleration.

fuzzard commented:
Jenkins build this please

wsnipex commented on tools/depends/target/gnutls/Makefile:
```diff
@@ -3,14 +3,10 @@ DEPS= ../../Makefile.include Makefile size-max.patch add-dl-as-private-lib.patch
 # lib name, version
 LIBNAME=gnutls
-VERSION=3.6.11.1
+VERSION=3.6.13
```
please bump to 3.6.14, since a serious security issue has been fixed: https://lists.gnupg.org/pipermail/gnutls-help/2020-June/004648.html
